src/search/mh_prog.cc

```diff
@@ -421,14 +435,29 @@ void mh_sampler::mcmc_iter(top_k_progs& topk_progs, int niter, prog* orig, bool
         delete curr;
         curr = new prog(*prog_start);
       }
-      cout << "start from program (error and performance costs: "
-           << prog_start->_error_cost << " " << prog_start-> _perf_cost << "):" << endl;
+      if(k2_config.functionality == FUNC_optimize){
+        cout << "start from program (error and performance costs: "
+            << prog_start->_error_cost << " " << prog_start-> _perf_cost << "):" << endl;
+      }else if (k2_config.functionality == FUNC_repair){
+        cout << "start from program (safety cost: "
+            << prog_start->_safety_cost << "):" << endl;
+        // todo: add print for error cost
+      }
       prog_start->print();
       if (is_win) { // reset validator original program
         inout_t::start_insn = win.first;
         inout_t::end_insn = win.second;
         init_sample_range(&prog_start->inst_list[win.first], (win.second - win.first + 1));
-        static_safety_check_pgm(prog_start->inst_list, inst::max_prog_len);
+        if(k2_config.functionality == FUNC_optimize){
+          // for repair this static check is not needed because original program may not be safe
+          if (static_safety_check_pgm(prog_start->inst_list, inst::max_prog_len, true) > 0){
+            // stop execution here 
+            // error message would have already been thrown by static_safety_check_pgm
+            // so throw empty err message to stop execution
+            string err_msg = "";
+            throw(err_msg);
+          }
+        }
```

Comment:
These checks and exception throwing seem pretty repetitive. Why do we need these? Could we get rid of them and put in something less repetitive and cleaner?